File: java/src/test/java/mrk/hackerrank/HackerRank06Test.java
package mrk.hackerrank;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

class HackerRank06Test {

    private static Stream<Arguments> testData() {
        return Stream.of(
                Arguments.of(new int[][]{{1, 2, 3}, {4, 5, 6}, {9, 8, 9}}, 2),
                Arguments.of(new int[][]{{1}}, 0),
                Arguments.of(new int[][]{{100, 0}, {0, 100}}, 200),
                Arguments.of(new int[][]{{-100, 0}, {0, -100}}, 200)
        );
    }

    @ParameterizedTest
    @MethodSource("testData")
    void test(int[][] arr, int expected) {
        Assertions.assertEquals(expected, new HackerRank06().diagonalDifference(arr));
    }

    @ParameterizedTest
    @MethodSource("invalidInputs")
    void testInvalidInput(int[][] arr) {
        Assertions.assertThrows(IllegalArgumentException.class, () -> new HackerRank06().diagonalDifference(arr));
    }

    private static Stream<Arguments> invalidInputs() {
        return Stream.of(
                Arguments.of((Object) new int[][]{{101, 0}, {0, 0}}),
                Arguments.of((Object) new int[][]{{0, 0}, {0, -101}}),
                Arguments.of((Object) new int[][]{{0, 200}, {0, 0}})
        );
    }
}

File: kotlin/src/test/kotlin/HackerRank06Test.kt
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

class HackerRank06Test {
    @ParameterizedTest(name = "{0}")
    @MethodSource("hackerRank06Data")
    fun test_hackerRank06(
        name: String,
        input: Array<IntArray>,
        output: Int,
    ) {
        assertEquals(output, hackerRank06(input))
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("invalidInputs")
    fun test_hackerRank06_invalidInput(
        name: String,
        input: Array<IntArray>,
    ) {
        assertThrows(IllegalArgumentException::class.java) { hackerRank06(input) }
    }

    companion object {
        @JvmStatic
        fun hackerRank06Data() =
            listOf(
                Arguments.of("3x3 matrix", arrayOf(intArrayOf(1, 2, 3), intArrayOf(4, 5, 6), intArrayOf(9, 8, 9)), 2),
                Arguments.of("1x1 matrix", arrayOf(intArrayOf(1)), 0),
                Arguments.of("boundary max", arrayOf(intArrayOf(100, 0), intArrayOf(0, 100)), 200),
                Arguments.of("boundary min", arrayOf(intArrayOf(-100, 0), intArrayOf(0, -100)), 200),
            )

        @JvmStatic
        fun invalidInputs() =
            listOf(
                Arguments.of("element above 100", arrayOf(intArrayOf(101, 0), intArrayOf(0, 0))),
                Arguments.of("element below -100", arrayOf(intArrayOf(0, 0), intArrayOf(0, -101))),
                Arguments.of("element far out of range", arrayOf(intArrayOf(0, 200), intArrayOf(0, 0))),
            )
    }
}
